### PR TITLE
RearrangeEpisodeGenerator

### PR DESCRIPTION
**Pull request by aclegg3**
## Motivation and Context

This PR adds the RearrangeEpisodeGenerator class for generating RearrangeEpisodes for a RearrangeDataset. 

Features include:
- Configuration:
  - Scene, object, and receptacle sets (w/ inclusion and exclusion constraints)
  - Samplers and parameters for scenes, clutter object placements, object placement targets, and ArticulatedObject joint states.
  - Markers (tracked points in the local space of objects and links)

- Samplers:
  - SingleScene and MultiScene
  - Object placements with rejection sampling, "snap to surface", and simulation settling to produce dynamically stable configurations. 
  - Target sampling to find valid re-configurations (same method as object sampling above) of previously sampled object instances without replacement.
  - ArticulatedObject joint states from configured ranges.
- Receptacles:
  - AABBs in object local space defined in SceneDataset configuration files "user_defined" fields.
  - Scraping receptacles from the object instanced in the scene and culling based on object and receptacle name substrings in receptacle sets.
Supporting Utilities (some of these will migrate into habitat_sim):
  - DebugVisualizer class for modular, easy production of images and videos during and after generator process.
  - Raycasting based "snap to surface" logic.

**Future Work**:
- migrate sim utilities to habitat_sim
- implement additional DebugVisualizer features for better debug visualization

## How Has This Been Tested

Note: testing this feature requires changes to ReplicaCAD ([v1.2](https://dl.fbaipublicfiles.com/habitat/ReplicaCAD/ReplicaCAD_dataset_v1.2.zip)) and YCB ([v1.1](https://dl.fbaipublicfiles.com/habitat/ycb/hab_ycb_v1.1.zip)) datasources. For simple setup, apply [this minor patch](https://github.com/facebookresearch/habitat-sim/commit/9eb06f2b15dc7718fde57c3c95acd417cd52b4ec) to habitat_sim and download the new assets with: 
`python habitat_sim/utils/datasets_download.py --uids rearrange_task_assets`
_These assets are not finalized, expect to re-download often until merged (**last updated 09/02**)._

**Running the generator:**
Running `python habitat/datasets/rearrange/rearrange_generator.py --run` 
will generate a single episode with default config and save it locally as `rearrange_ep_dataset.json.gz`.
Use:
- `--config` to pass a custom generator config file. **(See example config: habitat/datasets/rearrange/configs/test_config.yaml)**
- `--num-episodes` to specify the number of episodes to generate.
- `--out` to specify the output location for the serialized dataset.
- `--debug` to also generate and save debugging images/videos.
- `--db-output` to specify the output location for debugging images/videos (default is `rearrange_ep_gen_output/`).

**NOTE:** `python habitat/datasets/rearrange/rearrange_generator.py --list` will print concise list of scene, object, and receptacle handles for reference when authoring a custom generator config file.

Added a pytest: `pytest test/test_rearrange_task.py -k test_rearrange_episode_generator`
Set `debug_visualization=True` to produce videos of receptacles and final sampled simulation state for each episode. 

Example equirectangular image of YCB clutter objects generated in ReplicaCAD within a set of receptacles as produced by the generator test.
image

Example of 100 sampled placements of 1-30 objects within the fridge receptacles:

https://user-images.githubusercontent.com/1445143/131386888-e4250982-6806-40e6-80cc-937c36c1057a.mp4

## Types of changes

- Docs change / refactoring / dependency upgrade
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by aclegg3** [comment]
Tests are failing until we update the datasources for automated download ([Habitat-sim PR 1505](https://github.com/facebookresearch/habitat-sim/pull/1505))

File: habitat/datasets/rearrange/rearrange_generator.py [hunk 1]
```diff
@@ -0,0 +1,972 @@
...
+                    object_handles,
+                    receptacle_info,
+                    (
+                        obj_sampler_info["params"]["num_samples"][0],
+                        obj_sampler_info["params"]["num_samples"][1],
+                    ),
+                    obj_sampler_info["params"]["orientation_sampling"],
+                )
+            else:
+                print(
```
Comment (Skylion007) [inline habitat/datasets/rearrange/rearrange_generator.py:194]: Just set the error string.
```
                raise (NotImplementedError(f"Requested object sampler '{obj_sampler_info['type']}' is not implemented."))
```

File: habitat/datasets/rearrange/samplers.py [hunk 2]
```diff
@@ -0,0 +1,596 @@
...
+    def num_scenes(self) -> int:
+        return 1
+
+
+class MultiSceneSampler(SceneSampler):
+    """
+    Uniform sampling from a set of scenes.
+    """
+
+    def __init__(self, scenes: List[str]) -> None:
```
Comment (Skylion007) [inline habitat/datasets/rearrange/samplers.py:51]:
```suggestion
    def __init__(self, scenes: Sequence[str]) -> None:
```
This is actually the type you want.

File: habitat/datasets/rearrange/samplers.py [hunk 3]
```diff
@@ -0,0 +1,596 @@
...
+                        .object_id
+                    ]
+                snap_success = sutils.snap_down(
+                    sim,
+                    new_object,
+                    support_object_ids,
+                    vdb=vdb,
+                )
+                if snap_success:
+                    print(
```
Comment (Skylion007) [inline habitat/datasets/rearrange/samplers.py:256]: Use Logger

File: habitat/datasets/rearrange/samplers.py [hunk 4]
```diff
@@ -0,0 +1,596 @@
...
+                        f"Successfully sampled (snapped) object placement in {num_placement_tries} tries."
+                    )
+                    return new_object
+
+            elif not new_object.contact_test():
+                print(
```
Comment (Skylion007) [inline habitat/datasets/rearrange/samplers.py:262]: Use logger

File: habitat/datasets/rearrange/samplers.py [hunk 5]
```diff
@@ -0,0 +1,596 @@
...
+            num_pairing_tries += 1
+            new_object = self.single_sample(sim, snap_down, vdb)
+            if new_object is not None:
+                new_objects.append(new_object)
+
+        if len(new_objects) >= self.num_objects[0]:
+            return new_objects
+
+        # we didn't find the minimum number of placements, so remove all new objects and return
+        print(
```
Comment (Skylion007) [inline habitat/datasets/rearrange/samplers.py:328]: Please make all these use the logger.

**Review by Skylion007 — CHANGES_REQUESTED** [review]
Some initial refactoring is needed to remove all the print statements.

**Comment by aclegg3** [comment]
Encapsulated by #764.

**Pull request closed without merging**